# Amooora text data cleaning

### Import libraries and load csv

In [85]:
import pandas as pd

In [86]:
!tree

.
├── dedemenezes_exploratory.ipynb
├── okcupid_profiles_encoded.csv
├── raw_data
│   ├── okcupid_profiles.csv
│   ├── parsed_data_public.csv
│   ├── question_data.csv
│   └── user_data_public.csv
└── README.md

1 directory, 7 files


In [87]:
df = pd.read_csv('raw_data/okcupid_profiles.csv')

In [88]:
df.shape

(59946, 31)

In [89]:
# from sklearn.preprocessing import LabelEncoder


# categorical_cols = ['status', 'sex', 'orientation', 'body_type', 'diet', 'drinks', 'drugs',
#                     'education', 'ethnicity', 'job', 'location', 'smokes', 'religion', 'sign', 'pets']


# label_encoding_cols = ['education', 'ethnicity', 'job', 'location', 'sign']


# label_encoders = {}
# for col in label_encoding_cols:
#     le = LabelEncoder()
#     df[col] = le.fit_transform(df[col].astype(str))  # Convert to string to avoid NaN issues
#     label_encoders[col] = le  # Save the encoder in case you need to decode later


# df = pd.get_dummies(df, columns=[col for col in categorical_cols if col not in label_encoding_cols], drop_first=True)


# encoded_file_path = "okcupid_profiles_encoded.csv"  # Output file path
# df.to_csv(encoded_file_path, index=False)

# print(f"Dataset with encoding saved at: {encoded_file_path}")

In [90]:
df.head()

age     status sex orientation       body_type               diet  \
0   22     single   m    straight  a little extra  strictly anything   
1   35     single   m    straight         average       mostly other   
2   38  available   m    straight            thin           anything   
3   23     single   m    straight            thin         vegetarian   
4   29     single   m    straight        athletic                NaN   

     drinks      drugs                          education  \
0  socially      never      working on college/university   
1     often  sometimes              working on space camp   
2  socially        NaN     graduated from masters program   
3  socially        NaN      working on college/university   
4  socially      never  graduated from college/university   

             ethnicity  ...  \
0         asian, white  ...   
1                white  ...   
2                  NaN  ...   
3                white  ...   
4  asian, black, other  ...   

                                              essay0  \
0  about me:  i would love to think that i was so...   
1  i am a chef: this is what that means. 1. i am ...   
2  i'm not ashamed of much, but writing public te...   
3          i work in a library and go to school. . .   
4  hey how's it going? currently vague on the pro...   

                                              essay1  \
0  currently working as an international agent fo...   
1  dedicating everyday to being an unbelievable b...   
2  i make nerdy software for musicians, artists, ...   
3          reading things written by old dead people   
4                         work work work work + play   

                                              essay2  \
0  making people laugh. ranting about a good salt...   
1  being silly. having ridiculous amonts of fun w...   
2  improvising in different contexts. alternating...   
3  playing synthesizers and organizing books acco...   
4  creating imagery to look at: http://bagsbrown....   

                                              essay3  \
0  the way i look. i am a six foot half asian, ha...   
1                                                NaN   
2  my large jaw and large glasses are the physica...   
3                  socially awkward but i do my best   
4            i smile a lot and my inquisitive nature   

                                              essay4  \
0  books: absurdistan, the republic, of mice and ...   
1  i am die hard christopher moore fan. i don't r...   
2  okay this is where the cultural matrix gets so...   
3  bataille, celine, beckett. . . lynch, jarmusch...   
4  music: bands, rappers, musicians at the moment...   

                                              essay5  \
0                  food. water. cell phone. shelter.   
1  delicious porkness in all of its glories. my b...   
2  movement conversation creation contemplation t...   
3                                                NaN   
4                                                NaN   

                        essay6  \
0  duality and humorous things   
1                          NaN   
2                          NaN   
3   cats and german philosophy   
4                          NaN   

                                              essay7  \
0  trying to find someone to hang out with. i am ...   
1                                                NaN   
2  viewing. listening. dancing. talking. drinking...   
3                                                NaN   
4                                                NaN   

                                              essay8  \
0  i am new to california and looking for someone...   
1  i am very open and will share just about anyth...   
2  when i was five years old, i was known as "the...   
3                                                NaN   
4                                                NaN   

                                              essay9  
0  you want to be swept off your feet! you are ti...  
1    

In [91]:
text_df = df.iloc[:, 21:]

In [92]:
text_df.head()

essay0  \
0  about me:  i would love to think that i was so...   
1  i am a chef: this is what that means. 1. i am ...   
2  i'm not ashamed of much, but writing public te...   
3          i work in a library and go to school. . .   
4  hey how's it going? currently vague on the pro...   

                                              essay1  \
0  currently working as an international agent fo...   
1  dedicating everyday to being an unbelievable b...   
2  i make nerdy software for musicians, artists, ...   
3          reading things written by old dead people   
4                         work work work work + play   

                                              essay2  \
0  making people laugh. ranting about a good salt...   
1  being silly. having ridiculous amonts of fun w...   
2  improvising in different contexts. alternating...   
3  playing synthesizers and organizing books acco...   
4  creating imagery to look at: http://bagsbrown....   

                                              essay3  \
0  the way i look. i am a six foot half asian, ha...   
1                                                NaN   
2  my large jaw and large glasses are the physica...   
3                  socially awkward but i do my best   
4            i smile a lot and my inquisitive nature   

                                              essay4  \
0  books: absurdistan, the republic, of mice and ...   
1  i am die hard christopher moore fan. i don't r...   
2  okay this is where the cultural matrix gets so...   
3  bataille, celine, beckett. . . lynch, jarmusch...   
4  music: bands, rappers, musicians at the moment...   

                                              essay5  \
0                  food. water. cell phone. shelter.   
1  delicious porkness in all of its glories. my b...   
2  movement conversation creation contemplation t...   
3                                                NaN   
4                                                NaN   

                        essay6  \
0  duality and humorous things   
1                          NaN   
2                          NaN   
3   cats and german philosophy   
4                          NaN   

                                              essay7  \
0  trying to find someone to hang out with. i am ...   
1                                                NaN   
2  viewing. listening. dancing. talking. drinking...   
3                                                NaN   
4                                                NaN   

                                              essay8  \
0  i am new to california and looking for someone...   
1  i am very open and will share just about anyth...   
2  when i was five years old, i was known as "the...   
3                                                NaN   
4                                                NaN   

                                              essay9  
0  you want to be swept off your feet! you are ti...  
1                                                NaN  
2  you are bright, open, intense, silly, ironic, ...  
3                              you feel so inclined.  
4                                                NaN

### Cleaning dataset

In [93]:
text_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59946 entries, 0 to 59945
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   essay0  54458 non-null  object
 1   essay1  52374 non-null  object
 2   essay2  50308 non-null  object
 3   essay3  48470 non-null  object
 4   essay4  49409 non-null  object
 5   essay5  49096 non-null  object
 6   essay6  46175 non-null  object
 7   essay7  47495 non-null  object
 8   essay8  40721 non-null  object
 9   essay9  47343 non-null  object
dtypes: object(10)
memory usage: 4.6+ MB


#### Duplicates

In [94]:
text_df.duplicated().sum()

np.int64(2134)

In [95]:
text_df[text_df.duplicated()]

essay0 essay1 essay2 essay3 essay4 essay5 essay6 essay7 essay8 essay9
51       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
64       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
80       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
84       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
175      NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...
59757    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
59791    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
59857    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
59881    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
59930    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN

[2134 rows x 10 columns]

We have `2134`duplicated observations all made of empty answers to all questions

In [96]:
no_dup_df = text_df.drop_duplicates()
no_dup_df.shape

(57812, 10)

### Empty values

In [97]:
no_dup_df.isnull().sum().sort_values(ascending=False) / len(no_dup_df) * 100

essay8    29.564796
essay6    20.130769
essay9    18.108697
essay7    17.847506
essay3    16.161005
essay5    15.078184
essay4    14.536774
essay2    12.985193
essay1     9.409811
essay0     5.815402
dtype: float64

In [98]:
df.dropna().shape

(4407, 31)

In [99]:
df.shape

(59946, 31)

In [104]:
df.iloc[:, 0:21].isnull().sum() / len(df) * 100

age             0.000000
status          0.000000
sex             0.000000
orientation     0.000000
body_type       8.834618
diet           40.694959
drinks          4.979482
drugs          23.487806
education      11.056618
ethnicity       9.475194
height          0.005005
income          0.000000
job            13.675641
last_online     0.000000
location        0.000000
offspring      59.321723
pets           33.231575
religion       33.740366
sign           18.443266
smokes          9.194942
speaks          0.083408
dtype: float64

In [110]:
df.drugs.unique()

array(['never', 'sometimes', nan, 'often'], dtype=object)

In [109]:
df.diet.unique()

array(['strictly anything', 'mostly other', 'anything', 'vegetarian', nan,
       'mostly anything', 'mostly vegetarian', 'strictly vegan',
       'strictly vegetarian', 'mostly vegan', 'strictly other',
       'mostly halal', 'other', 'vegan', 'mostly kosher',
       'strictly halal', 'halal', 'strictly kosher', 'kosher'],
      dtype=object)

In [107]:
encoded_df = pd.read_csv('okcupid_profiles_encoded.csv')

In [108]:
encoded_df

age  education  ethnicity  height  income  job       last_online  \
0       22         25         98    75.0      -1   20  2012-06-28-20-30   
1       35         31        216    70.0   80000    8  2012-06-29-21-41   
2       38         12        189    68.0      -1   12  2012-06-27-09-10   
3       23         25        216    71.0   20000   19  2012-06-28-14-22   
4       29          9         22    66.0      -1    0  2012-06-27-21-26   
...    ...        ...        ...     ...     ...  ...               ...   
59941   59          9        189    62.0      -1   17  2012-06-12-21-47   
59942   24         25        217    72.0      -1    6  2012-06-29-11-01   
59943   42         12          0    71.0  100000    4  2012-06-27-23-37   
59944   27         25          1    73.0      -1   10  2012-06-23-13-01   
59945   39         12        216    68.0      -1   10  2012-06-29-00-42   

       location                               offspring  sign  ...  \
0           175  doesn't have kids, but might want them    16  ...   
1           118  doesn't have kids, but might want them     8  ...   
2           155                                     NaN    32  ...   
3            16                       doesn't want kids    29  ...   
4           155                                     NaN     0  ...   
...         ...                                     ...   ...  ...   
59941       118                                has kids    10  ...   
59942       155                       doesn't have kids    23  ...   
59943       175                       doesn't have kids    36  ...   
59944       155       doesn't have kids, but wants them    22  ...   
59945       155                                     NaN    18  ...   

      pets_has cats pets_has dogs pets_has dogs and dislikes cats  \
0             False         False                           False   
1             False         False                           False   
2              True         False                           False   
3             False         False                           False   
4             False         False                           False   
...             ...           ...                             ...   
59941         False          True                           False   
59942         False         False                           False   
59943         False         False                           False   
59944         False         False                           False   
59945         False         False                           False   

      pets_has dogs and has cats pets_has dogs and likes cats pets_likes cats  \
0                          False                        False           False   
1                          False                        False           False   
2                          False                        False           False   
3                          False                        False            True   
4                          False                        False           False   
...                          ...                          ...             ...   
59941                      False                        False           False   
59942                      False                        False           False   
59943                      False                        False           False   
59944                      False                        False           False   
59945                      False                        False           False   

      pets_likes dogs pets_likes dogs and dislikes cats  \
0               False                             False   
1               False                             False   
2               False                             False   
3               False                             False   
4               False                             False   
...               ...                               ...   
59941           False                             